Expand AEAD limits to consider multi-user security.

chris-wood opened the pull request:
In a typical multi-user setting, the adversary is assumed to perform some
(massive) amount of offline work to break the integrity of a single random
connection. Each user is represented as a unique (key, nonce) pair. In
considering the same threat model, we must treat each key resulting from
a KeyUpdate event as a unique user. As a result, any bounds implied from
the multi-user setting imply (minimally) for the lifetime of a single
connection.

Hoang et al. present tight multi-user security bounds for randomized AES-GCM
(as is used in TLS 1.3 and QUIC), so we can take advantage of those for
per-connection limits. In contrast, there are no multi-user security bounds
for AEAD_CHACHA20_POLY1305 or AEAD_AES_128_CCM, so we must use the
single-user bounds in their stead.

cc @martinthomson, @fxguenther

mimoo commented on draft-ietf-quic-transport.md:
```diff
@@ -1493,8 +1493,8 @@ supports.  The version field in the QUIC packet header is authenticated using
 be unambiguously different to ensure no confusion about their interpretation.
-One way that a new format could be introduced is to define a TLS extension with
-a different codepoint.
+When using TLS as the handshake protocol, defining a new TLS extension with a
```
two things:

* If we're really writing an agnostic-to-TLS document then I think we can remove this sentence all together.
* In order to be able to swap the QUIC-TLS doc, we should provide a clear API ( -> we pass these transport arguments, <- the security layer gives them back) and clear security properties for the transport layer (it should authenticate the transport parameters)

mimoo commented on draft-ietf-quic-transport.md:
```diff
@@ -2348,8 +2347,8 @@ have negotiated a packet protection scheme with a larger minimum AEAD expansion.
-valid QUIC packets.  With the set of AEAD functions defined in {{QUIC-TLS}},
-packets that are smaller than 21 bytes are never valid.
+valid QUIC packets.  When using TLS as the handshake protocol, no AEAD
+function defined in {{QUIC-TLS}} will yield packets smaller than 21 bytes.
```
again, not great. What should other ~transport layers~~ handshake protocol do? What

chris-wood replied:
We can add something that says "other handshake protocols MUST not yield packets less than 21 bytes." Thoughts?

mimoo commented on draft-ietf-quic-transport.md:
```diff
@@ -2575,9 +2574,11 @@ to have confidentiality and data origin authentication; the cryptographic
+to protect individual packets. After applying packet protection, packet numbers
+are themselves protected encryption. TLS-based packet number protection is
+described in see {{QUIC-TLS}} for details. Future versions may protect packet
+numbers in a different manner. The underlying packet number increases with
```
this one is tricky, as it is deeply entangled with the QUIC-TLS RFC

mimoo commented on draft-ietf-quic-transport.md:
```diff
@@ -2629,7 +2630,8 @@ Packet numbers are divided into 3 spaces in QUIC:
   {{packet-protected}} are in this space.
 
-As described in {{QUIC-TLS}}, each packet type uses different protection keys.
+Each packet type uses different protection keys; see {{QUIC-TLS}} for a description
```
As I understand, TLS gives us back different keys for different types of packets, and it's this RFC's job to encrypt stuff afterwards. If my understanding is correct, then we should probably list the type of keys here, and have a clear API as to how we retrieve these keys from the handshake protocol.

chris-wood replied:
No, not quite. Packet protection is described in the TLS doc, not here.

mimoo commented on draft-ietf-quic-transport.md:
this is good

mimoo commented on draft-ietf-quic-transport.md:
```diff
@@ -3354,18 +3360,20 @@ Long Packet Type (T):
   length of the packet number field is the value of this field, plus one.  These
-  bits are protected using header protection (see Section 5.4 of {{QUIC-TLS}}).
+  bits are protected using header protection (see Section 5.4 of {{QUIC-TLS}}
+  for an example using TLS).
```
I don't really like that TLS cannot be used as a blackbox and modifies part of the QUIC packets itself.

mimoo commented on draft-ietf-quic-transport.md:
```diff
@@ -5040,7 +5050,8 @@ CRYPTO_ERROR (0x1XX):
   Codes for errors occurring when TLS is used for the crypto handshake are
-  described in Section 4.8 of {{QUIC-TLS}}.
+  described in Section 4.8 of {{QUIC-TLS}}. Future QUIC versions SHOULD define
+  separate error CRYPTO_ERROR codes.
```
did you mean "separate CRYPTO_ERROR error codes" ?

chris-wood replied:
Yeah, will fix.

Commit:
Expand AEAD limits to consider multi-user security.

In a typical multi-user setting, the adversary is assumed to perform some
(massive) amount of offline work to break the integrity of a single random
connection. Each user is represented as a unique (key, nonce) pair. In
considering the same threat model, we must treat each key resulting from
a KeyUpdate event as a unique user. As a result, any bounds implied from
the multi-user setting imply (minimally) for the lifetime of a single
connection.

Hoang et al. present tight multi-user security bounds for randomized AES-GCM
(as is used in TLS 1.3 and QUIC), so we can take advantage of those for
per-connection limits. In contrast, there are no multi-user security bounds
for AEAD_CHACHA20_POLY1305 or AEAD_AES_128_CCM, so we must use the
single-user bounds in their stead.

Commit:
Expand AEAD limits to consider multi-user security.

In a typical multi-user setting, the adversary is assumed to perform some
(massive) amount of offline work to break the integrity of a single random
connection. Each user is represented as a unique (key, nonce) pair. In
considering the same threat model, we must treat each key resulting from
a KeyUpdate event as a unique user. As a result, any bounds implied from
the multi-user setting translate (minimally) to bounds for multiple keys
used in the course of a single connection. Note that this is not perfect,
as the total number of users in a standard multi-user setting is far larger
than the number of keys used in a single connection. However, absent realistic
"per-connection" bounds, this is the best we can do. (This is a gap in
the literature.)

Hoang et al. present tight multi-user security bounds for randomized AES-GCM
(as is used in TLS 1.3 and QUIC), so we can take advantage of those for
per-connection limits. In contrast, there are no multi-user security bounds
for AEAD_CHACHA20_POLY1305 or AEAD_AES_128_CCM, so we must use the
single-user bounds in their stead.

chris-wood commented:
Note: this is just a draft to get us started!

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -100,6 +100,15 @@ informative:
+      - ins: S. Tessaro
+      - ins: A. Thiruvengadam
+    date: 2018-10-18
+    target: https://dl.acm.org/doi/10.1145/3243734.3243816
```
You can cite this using `{{?GCM-MU=DOI.10.1145/3243734.3243816}}` inline.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1552,30 +1561,39 @@ number of attempts to forge packets. TLS achieves this by closing connections
-AEAD_AES_256_GCM, but this specification recommends a lower limit.  For
-AEAD_AES_128_CCM, the limit on the number of packets that fail authentication
-is 2^23.5; see {{ccm-bounds}}.
+: Due to the way that header protection protects the Key Phase, packets that
```
This looks like a rewrapping only.  Maybe revert that.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1552,30 +1561,39 @@ number of attempts to forge packets. TLS achieves this by closing connections
+If a key update is not possible, the endpoint MUST immediately close the
+connection. If the number of packets that fail authentication within the
+connection (across all keys) exceeds a limit that is specific to the AEAD in
+use, the endpoint MUST immediately close the connection. Applying these limits
```
An immediate close can still involve sending and receiving packets, so you need to add " and not attempt to process any further packets".  Or something along these lines.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -2170,10 +2188,10 @@ packet = 4cfe4189655e5cd55c41f69080575d7999c25a5bfb
+TLS {{?TLS13}} and {{AEBounds}} do not specify per-key or per-connection
+limits on usage for AEAD_AES_128_CCM. However, any AEAD that is used with QUIC
+requires limits on use that ensure that both confidentiality and integrity are
+preserved. This section documents that analysis.
```
Perhaps change to: "This section documents an analysis of per-key limits, which - absent an analysis of multi-user security - also set per-connection limits."

martinthomson reviewed (commented):
I have to admit that I don't really understand how this change is motivated.  How is use of disparate keys within the same connection any different than use of different keys in multiple connections?

Commit:
Remove per-key limit in favor of a single, per-connection limit.

Commit:
Revert wrapping.

Commit:
Remove MultiUserSecurity reference.

Commit:
Add back future relaxation notice.

chris-wood commented:
> How is use of disparate keys within the same connection any different than use of different keys in multiple connections?

It's not, really. As I understand it, we're accepting we don't have a global view of all users, and thus limit our view of "users" to "all keys in a given connection." As a result, it's possible that these limits are higher than we might expect in a typical (global) multi-user setting, but I'm not sure we can do much about that beyond bumping the block (and key) size.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1570,12 +1564,14 @@ Note:
+limit on number of packets that fail authentication is 2^93; see {{AEBounds}}.
+For AEAD_AES_128_CCM, the confidentiality limit is 2^23 encrypted packets, and
+the integrity limit is 2^23.5 forged packets; see {{ccm-bounds}}. Future
+analyses and specifications MAY relax these limits.
```
We probably need to split this paragraph now.  One for the text on key updates: update before you get to L[confidentiality].  And a separate one for stop using the connection when you get to L[integrity].

martinthomson replied:
How about:

```suggestion
Key updates MUST be initiated before confidentiality limits for the selected
AEAD are exceeded.  For AEAD_AES_128_GCM and AEAD_AES_256_GCM, the
confidentiality limit is 2^27 encrypted packets; see {{gcm-bounds}}.  For
AEAD_CHACHA20_POLY1305, the confidentiality limit is greater than the number of
possible packets (2^62) and so can be disregarded.  For AEAD_AES_128_CCM, the
confidentiality limit is 2^23 encrypted packets; see {{ccm-bounds}}.

A connection MUST be closed if the number of packets that fail authentication
exceed the integrity limit for the selected AEAD. For AEAD_AES_128_GCM, and the
integrity limit is 2^54 encrypted or forged packets; see {{gcm-bounds}}.  For
AEAD_AES_256_GCM, the integrity limit For AEAD_CHACHA20_POLY1305, the limit on
number of packets that fail authentication is 2^36; see {{AEBounds}}.  For
AEAD_AES_128_CCM, the integrity limit is 2^23.5 forged packets; see
{{ccm-bounds}}.

Future analyses and specifications MAY relax confidentiality or integrity limits
for an AEAD.
```

I changed the 2^93 to 2^36 for ChaCha.  I think that the 93 was in error.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -2220,6 +2208,66 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
+
+The goal is to limit this advantage to 2^-57, to match the target in
+{{?TLS13}}. In this inequality, the last term dominates the rest, so the others
+can be removed without affecting the result. This produces the approximation:
```
We should probably add a new limit for AEAD_AES_256_GCM.  That is, keep `k` in the following equation and then say 2^54 for 128-bit keys, and then "something much larger" for 256-bit keys, but setting a limit of 2^54 is probably good for that too.  (Honestly, if you have discarded that many packets and are still going, good for you.)

chris-wood replied:
Yeah, let's use the same limit for both, but show (in the appendix) the derivation of the k=256 case. Stand by for an update.

chris-wood replied:
Okay, done!

martinthomson reviewed (commented):
This is pretty good.  I know that the analysis in the appendix is going to cause some concerns, but it's better than what we previously had.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -2209,10 +2195,10 @@ v:
+The analyses that follow rely on a count of the number of block operations involved
+in producing each message. For simplicity, and to match the analysis of other AEAD
+functions in {{AEBounds}}, this analysis assumes a packet length of 2^10 blocks
+and a packet size limit of 2^14.
```
```suggestion
functions in {{AEBounds}}, this analysis assumes a packet length of 2^10 blocks;
that is, a packet size limit of 2^14 bytes.
```

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -2222,6 +2208,66 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
+
+- The amount of offline work done by an attacker does not dominate other factors
+in the analysis.
+
```
```suggestion

The bounds in {{?GCM-MU}} are tighter and more complete than those used in
{{AEBounds}}, which allows for larger limits than those described in {{?TLS13}}.
```

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -2229,6 +2229,8 @@ by v * l, the number of forgery attempts and the size of each packet (in blocks)
 
+The bounds in {{?GCM-MU}} are tighter and more complete than those used in
+{{AEBounds}}, which allows for larger limits than those described in {{?TLS13}}.
 ### Confidentiality Limit
```
```suggestion

### Confidentiality Limit
```
Just in case.

chris-wood replied:
Already done.

martinthomson reviewed (approved):
This should cause some consternation when we introduce it, but I don't think that people will get too badly steamed.  The math is sound, at least to the extent that the underlying analyses are sound :)  And it's pretty directly actionable.

chris-wood commented:
Thanks, @martinthomson! I'll draft an issue for the main repo and share it soon.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -2220,6 +2221,81 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
+v <= 2^182
+~~~
+
+This is substantially smaller than the limit for AEAD_AES_128_GCM. Thus,
```
```suggestion
This is substantially smaller than the limit for AEAD_AES_128_GCM. However,
```

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -2220,6 +2221,81 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
+~~~
+
+This is substantially smaller than the limit for AEAD_AES_128_GCM. Thus,
+for consistency, this document uses the larger of the two -- 2^54 -- for both.
```
```suggestion
this document recommends that the smaller limit from AEAD_AES_128_GCM also
apply to AEAD_AES_256_GCM.
```

(The only important bit being s/larger/smaller/.)

This is partly consistency (because it is easier to use a single number here) and partly being conservative (because that large number is just unbelievable), and partly because that smaller value is already large enough.

chris-wood replied:
Woof. That one was bad. Good catch!

martinthomson reviewed (approved):
I keep finding things, sorry.

chris-wood committed:
Update draft-ietf-quic-tls.md

Co-authored-by: Martin Thomson <mt@lowentropy.net>

chris-wood committed:
Update draft-ietf-quic-tls.md

Co-authored-by: Martin Thomson <mt@lowentropy.net>

ekr commented on draft-ietf-quic-tls.md:
```diff
@@ -1570,12 +1564,23 @@ Note:
+A connection MUST be closed if the number of packets that fail authentication
+exceed the integrity limit for the selected AEAD. For AEAD_AES_128_GCM, and the
+integrity limit is 2^54 encrypted or forged packets; see {{gcm-bounds}}.  For
+AEAD_AES_256_GCM, the integrity limit For AEAD_CHACHA20_POLY1305, the limit on
```
Something is wrong with this text.

chris-wood replied:
Fixed -- thanks!

ekr commented on draft-ietf-quic-tls.md:
```diff
@@ -2220,6 +2221,82 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
+v <= 2^182
+~~~
+
+This is substantially smaller than the limit for AEAD_AES_128_GCM. However,
```
Larger?

chris-wood replied:
Fixed.

Commit:
Merge branch 'caw/mu-limits-and-no-reset' of github.com:chris-wood/base-drafts into caw/mu-limits-and-no-reset

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1570,12 +1564,23 @@ Note:
-AEAD_AES_256_GCM, but this specification recommends a lower limit.  For
-AEAD_AES_128_CCM, the limit on the number of packets that fail authentication
-is 2^23.5; see {{ccm-bounds}}.
+A connection MUST be closed if the number of encrypted packets exceeds the
```
I missed this last time around.

I think that this is unrealistic.  People will make new connections, which have the same properties as new keys.  So we should just recommend key updates.

chris-wood replied:
Hmm, fair point. They are functionally the same thing since we're ignoring the number of users. That said, doesn't it seem inconsistent to drop connections in the case of integrity, but then re-key for confidentiality?

Commit:
Revert confidentiality limit to KeyUpdates.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1538,27 +1538,21 @@ discarded.
+distinguish the AEAD in use from a random permutation; see {{AEBounds}},
+{{ROBUST}}, and {{?GCM-MU=DOI.10.1145/3243734.3243816}}.
 
 Note:
```
I think that we can drop this note now that the failed authentications are counted globally.

chris-wood replied:
Good point -- dropped.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1586,13 +1591,13 @@ Note:
+specified for the number of packets that can be authenticated and for the
+number packets that can fail authentication. Providing a reference to any
+analysis upon which values are based - and any assumptions used in that
+analysis - allows limits to be adapted to varying usage conditions.
```
And maybe revert this change.

chris-wood replied:
Yep, done.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1538,27 +1538,34 @@ discarded.
 
 ## Minimum Key Update Frequency
 
-Key updates MUST be initiated before usage limits on packet protection keys are
-exceeded. For the cipher suites mentioned in this document, the limits in
-Section 5.5 of {{!TLS13}} apply. {{!TLS13}} does not specify a limit for
-AEAD_AES_128_CCM, but the analysis in {{ccm-bounds}} shows that a limit of 2^23
-packets can be used to obtain the same confidentiality protection as the limits
-specified in TLS.
```
I didn't get enough time to do this today, but what I think we need here is some of the framing from this.  And that likely comes with a new section title and maybe even some restructuring.

chris-wood replied:
I reorganized things a bit. There's a parallel flow for the connection and integrity limit text now that think works. @ekr, what do you think of the new text?

Commit:
Refactor, rename, and rejuggle.

Commit:
Revert one change.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1536,14 +1536,7 @@ After this period, old read keys and their corresponding secrets SHOULD be
-specified in TLS.
+## Limits on Key Usage
 
 The usage limits defined in TLS 1.3 exist for protection against attacks
```
This still jumps in a bit cold.

```suggestion
This document sets usage limits for AEAD keys to ensure that overuse of a single key does
not give an adversary a disproportionate advantage in attacking the confidentiality
and integrity of communications.

The usage limits defined in TLS 1.3 exist for protection against attacks
```

Or some such nonsense.

chris-wood replied:
I'm happy with that as an intro, though might tweak it a bit to emphasize that limits don't always apply to keys (but to connections).

martinthomson reviewed (commented):
I'm much happier with this.  Thanks for fixing this up Chris.

chris-wood committed:
Update draft-ietf-quic-tls.md

Co-authored-by: Martin Thomson <mt@lowentropy.net>

Commit:
Small tweak.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -1538,9 +1538,9 @@ discarded.
 
 ## Limits on Key Usage
```
Like the tweak.  What do you think about the heading?

```suggestion
## Limits on AEAD Usage
```

chris-wood replied:
I originally left it as "Limits on Key Usage" for consistency with RFC8446, though I guess that's not the best rationale. I'm fine with the suggestion.

chris-wood committed:
Update draft-ietf-quic-tls.md

Co-authored-by: Martin Thomson <mt@lowentropy.net>
